### PR TITLE
AA vote functionality

### LINKED ISSUE #7: Connect forms with backend
Connect form to backend as well as also discuss what the form consists of.

### LINKED ISSUE #19: Upvote Function
Create a JS function that increments the upvote counter for the marker object and the counter in the info window

### PR DESCRIPTION
**Pull request by kanbi18**

### CODE REVIEW

File: maptivist/src/main/webapp/index.html
Comment (flying-bus) [inline maptivist/src/main/webapp/index.html]: Maybe just turn this into type="submit" so the createMarker() callback gets called to draw the marker on the map?

I think you will want to coordinate with @TBarrett553 for this form.

The issue is that the form submit sends a redirect, which causes the map to reload and the markers are no longer visible on the map. But if the createMarker() uses the fetch API to send the POST in the background, that won't cause a redirect.

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js]: I suggest creating a new servelet to handle upvotes ("/upvote"). Since that's a lot more work, perhaps its better to do that is a separate PR?

**Review by flying-bus — CHANGES_REQUESTED** [review]

**Review by ayomide321 — APPROVED** [review]

**Review by TBarrett553 — APPROVED** [review]

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java]: Maybe I missed it? But I don't see where the "title" parameter gets set in the HTML form?

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java]: This creates a single voteCount that's shared by all markers? I don't think that's correct?

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java]: What if two markers have the same title?

File: maptivist/src/main/webapp/index.html
Comment (TBarrett553) [reply maptivist/src/main/webapp/index.html]: Yeah for right now, the button works and calls the marker. However, when the form tries to submit there is a null pointer error in the master. But if you tested submitting the form with the improved marker servlet in this PR, I think the problem will be fixed.

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (TBarrett553) [reply maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java]: I think it's technically marker-title

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: I thought each Marker was going to have its own set of flags? Why is this a global then?

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: PR #44 moves these functions into the Marker class (and adds unit tests, etc). I suggest waiting until #44 is merged into master and using the Marker version.

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java:61]: After #44, there will be an ID field as well. It's a better idea to use the ID isntead of the title, in case multiple markers have the same title.

I suggest waiting until #44 is merged and then switching to the ID.

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: Shouldn't this be accessing the Datastore to get the actual Marker out and returning its flags?

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java]: Again, not a fan of using the a global here.

Instead we should be fetching the Marker from the Datastore and returning its votes here.

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js]: This one doesn't call getVote()?

**Review by flying-bus — COMMENTED** [review]
I noticed that there will be a lot of conflicts with #44.

I also think that the ID that is added in #44 would be useful - plus it simplifies the Marker <-> Entity conversion, so should also make this code easier.

Can we wait until #44 is merged and then I'll re-review this one. (I've already approved #44, so if you can give it a look then Teanna can merge it).

File: maptivist/src/main/java/com/google/sps/Marker.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/Marker.java:128]: Why is this deleted?

File: maptivist/src/main/java/com/google/sps/Marker.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/Marker.java:227]: Why is this empty? Shouldn't it be:
`createFlagString(this.flags)`

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java]: Ok - let's talk more about using a static/global for this.

Think about what happens if two visitors of your website add a flag to different markers at the same time. Assume one of them is on a slow connection and the other is on a fast connection.

Slow user upvotes Marker "A", which has a vote total of 100. They click the submit button which calls POST, and sets voteCount to 101. The doPost sends the redirect to index.html, and the slow user reloads the page.

While the slow user is reloading, the fast user upvotes Marker "B" which has a vote total of 0. They click the submit button which calls POST and sets voteCount to 1. The doPost sends the redirect to index.html. Because they are on a fast connection, the page reload happens quickly. The reload triggers the GET request to show the votes, and it retrieves the value 1 and shows it on the page.

Now the slow user's page finally sends the GET request to see the votes. The value of voteCount is 1, because that's what the fast user's POST set it to. It retrieves the value 1, and puts it on the page.  So now the slow user say votes = 100, clicked upvote and now sees votes = 1... So somehow they deleted votes?

Like I said earlier, we should avoid using a static/global variable to store the votes. And instead, GET should retrieve the value from the datastore.

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js]: Is this ID stable?

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: See below comment on global/static variables.

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: This looks identital to the version in VoteServlet.

We should have only a single definition and share it. Perhaps make it a static function of the Marker class?

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js:27]: delete extra newlines.

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js:46]: extra newlines?

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js]: If you did getElementById("vote-button") you could get the ID and pass it to the GET request as a parameter. That way you could look up the entity by id from GET.

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: Shouldn't we also look up by id here instead of title?

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: Why not just create a Marker from the Entity using Teanna's improvements?
```
Marker marker(getEntity(datastore, titleMatch));
ArrayList<string> flags = marker.flags();
```

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: And then here:

```
marker.addFlagString(flag);
datastore.put(marker.toEntity());
```

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java:70]: This version isnt used anymore, so we can delete it.

File: maptivist/src/main/java/com/google/sps/servlets/MarkerServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/MarkerServlet.java:69]: Why is this commented out?

File: maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/FlagServlet.java]: doGet still uses this static variable.

File: maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java
Comment (flying-bus) [inline maptivist/src/main/java/com/google/sps/servlets/VoteServlet.java:50]: On one hand, I applaud the code sharing.

On the other hand, I think its better to have separate GET and POST methods. For a typical server's API design, a GET will never change the server state. A POST usually changes server state (though not always).

So it makes sense to implement GET "/votes" that just returns the current vote count and implement a POST "/votes" that updates the votes. That way, in your HTML/Javascript, if you only want to get the current vote count, it's obvious by only sending a GET. If you want to update the votes, then it's obvious by sending a POST.

The way you have it works, and it has lots of nice code re-use :) so if you just want to merge it in like this, that's fine. But here's an optional suggestion:


1) Rename doPost to something like: `getAndMaybeUpdateVotes(bool update, HttpServletRequest request, HttpServeletResponse)`

2) Add:
```
public void doGet(HttpServeletRequest request, HttpServeletResponse) {
  getAndMaybeUpdateVotes(/* update= */ false, request, response);
}

public void doPost(HttpServeletRequest request, HttpServeletResponse) {
  getAndMaybeUpdateVotes( /* update= */ true, request, response);
}
```

File: maptivist/src/main/webapp/script.js
Comment (flying-bus) [inline maptivist/src/main/webapp/script.js:249]: If you implement the servlet like I suggested, then just change `method: "POST"` to `"GET"` and you can drop the `params.append("update", false)` line.

**Review by flying-bus — APPROVED** [review]
A little clean up left to do in the flags servlet - but since it's not hooked up yet I'm ok if you leave it to later (just make sure the github issue stays open).

Otherwise, just a few clean up comments and one OPTIONAL suggestion. But otherwise looks good.

File: maptivist/src/main/webapp/script.js
Comment (TBarrett553) [inline maptivist/src/main/webapp/script.js]: Thanks for that change!

**Pull request merged**
